src/key.rs

```diff
@@ -0,0 +1,309 @@
+        writeln!(f, "{:?} {:?}:", self.get_type().unwrap(), self.get_length().unwrap())?;
+        write!(f, "{}", s)
+    }
+}
+
+impl Drop for Key {
+    fn drop(&mut self)
+    {
+        unsafe {
+            lib::yaca_key_destroy(self.handle as *mut c_void)
+        }
+    }
+}
+
+impl Key {
+    /// Generates a secure key or key generation parameters
+    /// (or an Initialization Vector).
+    pub fn generate(key_type: &KeyType, key_length: &KeyLength) -> Result<Key>
+    {
+        key_generate(key_type, key_length)
+    }
+    /// Generates a secure private asymmetric key from parameters.
+    pub fn generate_from_parameters(params: &Key) -> Result<Key>
+    {
+        key_generate_from_parameters(params)
+    }
+    /// Imports a key or key generation parameters.
+    pub fn import(data: &[u8], key_type: &KeyType, password: Option<&CStr>) -> Result<Key>
+    {
+        key_import(data, key_type, password)
+    }
+    /// Exports a key or key generation parameters to arbitrary format.
+    pub fn export(&self, key_fmt: &KeyFormat, key_file_fmt: &KeyFileFormat,
+                  password: Option<&CStr>) -> Result<Vec<u8>>
+    {
+        key_export(self, key_fmt, key_file_fmt, password)
+    }
+    /// Extracts public key from a private one.
+    pub fn extract_public(&self) -> Result<Key>
+    {
+        key_extract_public(self)
+    }
+    /// Extracts parameters from a private or a public key.
+    pub fn extract_parameters(&self) -> Result<Key>
+    {
+        key_extract_parameters(self)
+    }
+    /// Derives a shared secret using Diffie-Helmann or EC Diffie-Helmann
+    /// key exchange protocol.
+    pub fn derive_dh(prv_key: &Key, pub_key: &Key) -> Result<Vec<u8>>
+    {
+        key_derive_dh(prv_key, pub_key)
+    }
+    /// Derives a key material from shared secret.
+    pub fn derive_kdf(kdf: &Kdf, algo: &DigestAlgorithm,
+                      secret: &[u8], info: Option<&[u8]>,
+                      key_material_len: usize) -> Result<Vec<u8>>
+    {
+        key_derive_kdf(kdf, algo, secret, info, key_material_len)
+    }
+    /// Derives a key from user password (PKCS #5 a.k.a. pbkdf2 algorithm).
+    pub fn derive_pbkdf2(password: &CStr, salt: Option<&[u8]>, iterations: usize,
+                         algo: &DigestAlgorithm, key_bit_len: usize) -> Result<Key>
+    {
+        key_derive_pbkdf2(password, salt, iterations, algo, key_bit_len)
+    }
+    /// Gets key's type.
+    pub fn get_type(&self) -> Result<KeyType>
+    {
+        key_get_type(&self)
+    }
+    /// Gets key's length.
+    pub fn get_length(&self) -> Result<KeyLength>
+    {
+        key_get_length(&self)
+    }
+}
+
+// Used by the C wrappers
+#[inline]
+pub(crate) fn get_handle(key: &Key) -> *const c_void
+{
+    key.handle
+}
+
+// Used to compose Keys in Seal
+#[inline]
+pub(crate) fn new_key(handle: *const c_void) -> Key
+{
+    Key{handle}
+}
+
+
+#[inline]
+fn key_get_type(key: &Key) -> Result<KeyType>
+{
+    let mut kt = -1;
```

Comment:
I wasn't going to trust 0 in get_length tbh. Key length should be >=8 and divisible by 8. What do you suggest here? Looking at this now, I'd change key_length_c_to_rs() to accept only >=8, but this would cause panic in both those functions, so I guess it's not what you want.

Reply:
Assertions seem preferable most of the time so any solution that uses them instead of panics might be desirable. After all, ```yaca``` we trust. An assertion triggering just means an error in ```yaca-rust``` (as in, misuse of ```yaca```), not an unavoidable external error condition.

Reply:
Assertions are always checked in both debug and release builds, and cannot be disabled. See debug_assert! for assertions that are not enabled in release builds by default.

:-D

s/assert/debug_assert (excl tests)

Reply:
No panics anywhere (there is one case that would fit panic in tests, but I used assert, just to be consistent).
src use debug_assert
tests use assert

Reply:
On the other hand I'm not sure this is a good choice. On one hand we trust yaca. On the other we should *guarantee* safety of safe code in Rust (like you argued with me in one case). Asserts are an accepted way to do that. This is not a performance critical code so we don't loose much by having asserts in case of things like handle and incorrect enum values.

Thoughts?

Reply:
There's no way sure-fire way to guarantee safety if we don't trust FFI callees. One invalid non-null ctx handle and voila.

Reply:
Yeah, I though about that myself after I wrote that comment. Ok, leaving debug_asserts.